tidy3d/plugins/resonance/resonance.py

```diff
@@ -0,0 +1,329 @@
+            raise ValidationError("The value of fmax is less than the value of fmin.")
+        return val
+
+    @validator("num_freqs_bounds", always=True)
+    def _check_num_freqs_bounds(cls, val):
+        """Validate freq_window"""
+        if val[1] < val[0]:
+            raise ValidationError("The value of nmax is less than the value of nmin.")
+        return val
+
+    def run(self, signal: ArrayLike[complex, 1], time_step: float) -> xr.Dataset:
+        """Finds resonances in a time series.
+        Note that the signal should start after the sources have turned off.
+
+        Parameters
+        ----------
+        signal : ArrayLike[complex, 1]
+            Time series to search for resonances
+        time_step : float
+            Time step / sampling rate of the data (in seconds).
+
+        Returns
+        -------
+        :class:`.xr.Dataset`
+            Dataset containing the frequency, decay rate, Q, amplitude, phase,
+            and estimation error of the resonances. Modes with low Q, small
+            amplitude, or high estimation error are likely to be spurious.
+        """
+        fmin, fmax = self.freq_window
+        nmin, nmax = self.num_freqs_bounds
+        nfreqs = int(time_step * len(signal) * (fmax - fmin) / 2)
+        nfreqs = min(nmax, nfreqs)
+        nfreqs = max(nmin, nfreqs)
+        if nfreqs == 0:
+            raise SetupError("Attempting to start with no frequencies.")
+        log.info(f"\tRunning ResonanceFinder (nfreqs = {nfreqs})")
+        omegas = np.linspace(
+            time_step * fmin * 2 * np.pi,
+            time_step * fmax * 2 * np.pi,
+            nfreqs,
+        )
+        eigvals = np.exp(-1j * omegas)
+        prev_num_eigvals = -1
+        for _ in range(1 + 2 * nfreqs):
+            resdata = self._iterate(signal, eigvals)
+            new_num_eigvals = len(resdata.eigvals)
+            if new_num_eigvals == prev_num_eigvals:
+                log.info(f"\tCompleted ResonanceFinder (nfreqs = {new_num_eigvals})")
+                break
+            log.info(f"\tIterated ResonanceFinder (nfreqs = {new_num_eigvals})")
+            prev_num_eigvals = new_num_eigvals
+
+        resonances = self._get_resonance_info(data=resdata, time_step=time_step)
+
+        return resonances.sortby("freq")
+
+    def run_scalar_field_time(self, signal: ScalarFieldTimeDataArray) -> xr.Dataset:
+        """Finds resonances in a 'ScalarFieldTimeDataArray'.
+        The time coordinates in the supplied 'ScalarFieldTimeDataArray' must be
+        uniformly spaced to use the resonance finder. The time step is calculated automatically.
+        Note that the signal should start after the sources have turned off.
+
+        Parameters
+        ----------
+        signal : ScalarFieldTimeDataArray
+            Time series to search for resonances
+
+        Returns
+        -------
+        :class:`.xr.Dataset`
+            Dataset containing the frequency, decay rate, Q, amplitude, phase,
```

Comment:
lets say here that the frequency is a coordinate.
"Dataset containing the decay rate, Q, amplitude, phase, and estimation error of the resonances as a function of frequency."